tensorboard/plugins/pr_curve/pr_curves_plugin.py

```diff
@@ -0,0 +1,155 @@
+    """Instantiates a PrCurvesPlugin.
+    Args:
+      context: A base_plugin.TBContext instance. A magic container that
+        TensorBoard uses to make objects available to the plugin.
+    """
+    self._multiplexer = context.multiplexer
+
+  @wrappers.Request.application
+  def pr_curves_route(self, request):
+    """A route that returns a JSON mapping between runs and PR curve data.
+
+    Returns:
+      Given a tag and a comma-separated list of runs (both stored within GET
+      parameters), fetches a JSON object that maps between run name and objects
+      containing data required for PR curves for that run. Runs that either
+      cannot be found or that lack tags will be excluded from the response.
+    """
+    runs = request.args.getlist('run')
+    tag = request.args.get('tag')
+
+    response_mapping = {}
+    for run in runs:
+      try:
+        tensor_events = self._multiplexer.Tensors(run, tag)
+      except KeyError:
+        return http_util.Respond(
+            request,
+            'No PR curves could be fetched for run %r and tag %r' % (run, tag),
+            'text/plain',
+            400)
+      
+      response_mapping[run] = [
+          self._process_tensor_event(e) for e in tensor_events]
+
+    # We convert the tensor data into a JSON-able response.
+    return http_util.Respond(request, response_mapping, 'application/json')
+
+  @wrappers.Request.application
+  def tags_route(self, request):
+    """A route (HTTP handler) that returns a response with tags.
+    Returns:
+      A response that contains a JSON object. The keys of the object
+      are all the runs. Each run is mapped to a (potentially empty)
+      list of all tags that are relevant to this plugin.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    response = {
+        run: list(tag_to_content.keys())
+             for (run, tag_to_content) in all_runs.items()
+    }
+    return http_util.Respond(request, response, 'application/json')
+
+  @wrappers.Request.application
+  def steps_per_run_route(self, request):
+    """Gets a dict mapping run to a list of numeric steps.
+
+    Returns:
+      A dict with string keys (all runs with PR curve data). The values of the
+      dict are lists of step values (ints) that should be used for the slider
+      for that run.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    response = {}
+
+    # Compute the max step exhibited by any tag for each run. If a run lacks
+    # data, exclude it from the mapping.
+    for run, tag_to_content in all_runs.items():
+      if not tag_to_content:
+        # This run lacks data for this plugin.
+        continue
+
+      # Just use the list of tensor events for any of the tags to determine the
+      # steps to list for the run. The steps should be the same across tags
+      # unless the user uses conditionals the only write summaries now and then.
```

Comment:
Wait, isn't this really common? People rarely write summaries at every step in practice, because it would slow down training and the extra granularity isn't useful.

What _should_ be the correct behavior in the case where the steps are not the same across tags?